// @polymarket/clob-client-v2/dist/order-builder/helpers/buildOrderCreationArgs.cjs
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.cjs');
const require_constants = require('../../constants.cjs');
const require_config = require('../../config.cjs');
const require_getOrderRawAmounts = require('./getOrderRawAmounts.cjs');
let viem = require("viem");

//#region src/order-builder/helpers/buildOrderCreationArgs.ts
async function buildOrderCreationArgs(signer, maker, signatureType, userOrder, roundConfig, version = 2) {
	const { side, rawMakerAmt, rawTakerAmt } = require_getOrderRawAmounts.getOrderRawAmounts(userOrder.side, userOrder.size, userOrder.price, roundConfig);
	const makerAmount = (0, viem.parseUnits)(rawMakerAmt.toString(), require_config.COLLATERAL_TOKEN_DECIMALS).toString();
	const takerAmount = (0, viem.parseUnits)(rawTakerAmt.toString(), require_config.COLLATERAL_TOKEN_DECIMALS).toString();
	if (version === 1) {
		const v1Order = userOrder;
		return {
			maker,
			taker: v1Order.taker ?? viem.zeroAddress,
			tokenId: userOrder.tokenID,
			makerAmount,
			takerAmount,
			side,
			signer,
			signatureType,
			feeRateBps: v1Order.feeRateBps?.toString() ?? "0",
			nonce: v1Order.nonce?.toString() ?? "0",
			expiration: userOrder.expiration !== void 0 ? userOrder.expiration.toString() : "0"
		};
	}
	return {
		maker,
		tokenId: userOrder.tokenID,
		makerAmount,
		takerAmount,
		side,
		signer,
		signatureType,
		timestamp: Date.now().toString(),
		metadata: "metadata" in userOrder ? userOrder.metadata ?? require_constants.bytes32Zero : require_constants.bytes32Zero,
		builder: userOrder.builderCode ?? require_constants.bytes32Zero,
		expiration: userOrder.expiration !== void 0 ? userOrder.expiration.toString() : "0"
	};
}

//#endregion
exports.buildOrderCreationArgs = buildOrderCreationArgs;
