// @polymarket/clob-client-v2/dist/order-builder/helpers/buildMarketOrderCreationArgs.cjs
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.cjs');
const require_constants = require('../../constants.cjs');
const require_config = require('../../config.cjs');
const require_getMarketOrderRawAmounts = require('./getMarketOrderRawAmounts.cjs');
let viem = require("viem");

//#region src/order-builder/helpers/buildMarketOrderCreationArgs.ts
async function buildMarketOrderCreationArgs(signer, maker, signatureType, userMarketOrder, roundConfig, version = 2) {
	const { side, rawMakerAmt, rawTakerAmt } = require_getMarketOrderRawAmounts.getMarketOrderRawAmounts(userMarketOrder.side, userMarketOrder.amount, userMarketOrder.price || 1, roundConfig);
	const makerAmount = (0, viem.parseUnits)(rawMakerAmt.toString(), require_config.COLLATERAL_TOKEN_DECIMALS).toString();
	const takerAmount = (0, viem.parseUnits)(rawTakerAmt.toString(), require_config.COLLATERAL_TOKEN_DECIMALS).toString();
	if (version === 1) {
		const v1Order = userMarketOrder;
		return {
			maker,
			taker: v1Order.taker ?? viem.zeroAddress,
			tokenId: userMarketOrder.tokenID,
			makerAmount,
			takerAmount,
			side,
			signer,
			signatureType,
			feeRateBps: v1Order.feeRateBps?.toString() ?? "0",
			nonce: v1Order.nonce?.toString() ?? "0"
		};
	}
	return {
		maker,
		tokenId: userMarketOrder.tokenID,
		makerAmount,
		takerAmount,
		side,
		signer,
		signatureType,
		timestamp: Date.now().toString(),
		metadata: "metadata" in userMarketOrder ? userMarketOrder.metadata ?? require_constants.bytes32Zero : require_constants.bytes32Zero,
		builder: userMarketOrder.builderCode ?? require_constants.bytes32Zero
	};
}

//#endregion
exports.buildMarketOrderCreationArgs = buildMarketOrderCreationArgs;
