// @polymarket/clob-client-v2/dist/order-builder/orderBuilder.cjs
const require_signatureTypeV2 = require('../order-utils/model/signatureTypeV2.cjs');
const require_createMarketOrder = require('./helpers/createMarketOrder.cjs');
const require_createOrder = require('./helpers/createOrder.cjs');

//#region src/order-builder/orderBuilder.ts
var OrderBuilder = class {
	signer;
	chainId;
	signatureType;
	funderAddress;
	/**
	* Optional function to dynamically resolve the signer.
	* If provided, this function will be called to obtain a fresh signer instance
	* (e.g., for smart contract wallets or when the signer may change).
	* Should return a Wallet or JsonRpcSigner, or a Promise resolving to one.
	* If not provided, the static `signer` property is used.
	*/
	getSigner;
	constructor(signer, chainId, signatureType, funderAddress, getSigner) {
		this.signer = signer;
		this.chainId = chainId;
		this.signatureType = signatureType === void 0 ? require_signatureTypeV2.SignatureTypeV2.EOA : signatureType;
		this.funderAddress = funderAddress;
		this.getSigner = getSigner;
	}
	/**
	* Generate and sign a order
	*/
	async buildOrder(userOrder, options, version) {
		return require_createOrder.createOrder(await this.resolveSigner(), this.chainId, this.signatureType, this.funderAddress, userOrder, options, version);
	}
	/**
	* Generate and sign a market order
	*/
	async buildMarketOrder(userMarketOrder, options, version) {
		return require_createMarketOrder.createMarketOrder(await this.resolveSigner(), this.chainId, this.signatureType, this.funderAddress, userMarketOrder, options, version);
	}
	/** Unified getter: use fresh signer if available */
	async resolveSigner() {
		if (this.getSigner) {
			const s = await this.getSigner();
			if (!s) throw new Error("getSigner() function returned undefined or null");
			return s;
		}
		return this.signer;
	}
};

//#endregion
exports.OrderBuilder = OrderBuilder;
