// @polymarket/clob-client-v2/dist/types/clob.cjs
const require_side = require('../order-utils/model/side.cjs');

//#region src/types/clob.ts
let OrderType = /* @__PURE__ */ function(OrderType$1) {
	OrderType$1["GTC"] = "GTC";
	OrderType$1["FOK"] = "FOK";
	OrderType$1["GTD"] = "GTD";
	OrderType$1["FAK"] = "FAK";
	return OrderType$1;
}({});
let Chain = /* @__PURE__ */ function(Chain$1) {
	Chain$1[Chain$1["POLYGON"] = 137] = "POLYGON";
	Chain$1[Chain$1["AMOY"] = 80002] = "AMOY";
	return Chain$1;
}({});
let PriceHistoryInterval = /* @__PURE__ */ function(PriceHistoryInterval$1) {
	PriceHistoryInterval$1["MAX"] = "max";
	PriceHistoryInterval$1["ONE_WEEK"] = "1w";
	PriceHistoryInterval$1["ONE_DAY"] = "1d";
	PriceHistoryInterval$1["SIX_HOURS"] = "6h";
	PriceHistoryInterval$1["ONE_HOUR"] = "1h";
	return PriceHistoryInterval$1;
}({});
let AssetType = /* @__PURE__ */ function(AssetType$1) {
	AssetType$1["COLLATERAL"] = "COLLATERAL";
	AssetType$1["CONDITIONAL"] = "CONDITIONAL";
	return AssetType$1;
}({});

//#endregion
exports.AssetType = AssetType;
exports.Chain = Chain;
exports.OrderType = OrderType;
exports.PriceHistoryInterval = PriceHistoryInterval;
