// @polymarket/clob-client-v2/dist/http-helpers/index.cjs
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.cjs');
let axios = require("axios");
axios = require_rolldown_runtime.__toESM(axios);
let browser_or_node = require("browser-or-node");

//#region src/http-helpers/index.ts
const GET = "GET";
const POST = "POST";
const DELETE = "DELETE";
const overloadHeaders = (method, headers = {}) => {
	if (browser_or_node.isBrowser) return;
	headers["User-Agent"] = `@polymarket/clob-client`;
	headers.Accept = "*/*";
	headers.Connection = "keep-alive";
	headers["Content-Type"] = "application/json";
	if (method === GET) headers["Accept-Encoding"] = "gzip";
};
const request = async (endpoint, method, headers, data, params) => {
	overloadHeaders(method, headers);
	return await (0, axios.default)({
		method,
		url: endpoint,
		headers,
		data,
		params
	});
};
const sleep = (ms) => new Promise((res) => setTimeout(res, ms));
const post = async (endpoint, options, retryOnError = false) => {
	try {
		return (await request(endpoint, POST, options?.headers, options?.data, options?.params)).data;
	} catch (err) {
		if (retryOnError && isTransientAxiosError(err)) {
			await sleep(30);
			try {
				return (await request(endpoint, POST, options?.headers, options?.data, options?.params)).data;
			} catch (retryErr) {
				return errorHandling(retryErr);
			}
		}
		return errorHandling(err);
	}
};
const get = async (endpoint, options) => {
	try {
		return (await request(endpoint, GET, options?.headers, options?.data, options?.params)).data;
	} catch (err) {
		return errorHandling(err);
	}
};
const del = async (endpoint, options) => {
	try {
		return (await request(endpoint, DELETE, options?.headers, options?.data, options?.params)).data;
	} catch (err) {
		return errorHandling(err);
	}
};
const errorHandling = (err) => {
	if (axios.default.isAxiosError(err)) {
		if (err.response) {
			if (err.response?.data) {
				if (typeof err.response?.data === "string" || err.response?.data instanceof String) return {
					error: err.response?.data,
					status: err.response?.status
				};
				if (!Object.prototype.hasOwnProperty.call(err.response?.data, "error")) return {
					error: err.response?.data,
					status: err.response?.status
				};
				return {
					...err.response?.data,
					status: err.response?.status
				};
			}
		}
		if (err.message) return { error: err.message };
	}
	return { error: err };
};
const isTransientAxiosError = (err) => {
	if (!axios.default.isAxiosError(err)) return false;
	if (!err.response) return true;
	const status = err.response.status ?? 0;
	if (status >= 500 && status < 600) return true;
	const code = (err.code ?? "").toString();
	return [
		"ECONNABORTED",
		"ENETUNREACH",
		"EAI_AGAIN",
		"ETIMEDOUT"
	].includes(code);
};
const parseDropNotificationParams = (dropNotificationParams) => {
	const params = {};
	if (dropNotificationParams !== void 0) {
		if (dropNotificationParams.ids !== void 0) params.ids = dropNotificationParams?.ids.join(",");
	}
	return params;
};

//#endregion
exports.DELETE = DELETE;
exports.GET = GET;
exports.POST = POST;
exports.del = del;
exports.get = get;
exports.parseDropNotificationParams = parseDropNotificationParams;
exports.post = post;
