// @polymarket/clob-client-v2/dist/signing/eip712.cjs
const require_constants = require('./constants.cjs');
const require_signer = require('./signer.cjs');

//#region src/signing/eip712.ts
/**
* Builds the canonical Polymarket CLOB EIP712 signature
* @param signer
* @param ts
* @returns string
*/
const buildClobEip712Signature = async (signer, chainId, timestamp, nonce, address) => {
	const resolvedAddress = address ?? await require_signer.getSignerAddress(signer);
	const ts = timestamp.toString();
	return await require_signer.signTypedDataWithSigner({
		signer,
		domain: {
			name: "ClobAuthDomain",
			version: "1",
			chainId
		},
		types: { ClobAuth: [
			{
				name: "address",
				type: "address"
			},
			{
				name: "timestamp",
				type: "string"
			},
			{
				name: "nonce",
				type: "uint256"
			},
			{
				name: "message",
				type: "string"
			}
		] },
		value: {
			address: resolvedAddress,
			timestamp: ts,
			nonce,
			message: require_constants.MSG_TO_SIGN
		},
		primaryType: "ClobAuth"
	});
};

//#endregion
exports.buildClobEip712Signature = buildClobEip712Signature;
