// @polymarket/clob-client-v2/dist/endpoints.js
//#region src/endpoints.ts
const OK = "/ok";
const HEARTBEAT = "/v1/heartbeats";
const TIME = "/time";
const CREATE_API_KEY = "/auth/api-key";
const GET_API_KEYS = "/auth/api-keys";
const DELETE_API_KEY = "/auth/api-key";
const DERIVE_API_KEY = "/auth/derive-api-key";
const CLOSED_ONLY = "/auth/ban-status/closed-only";
const GET_SAMPLING_SIMPLIFIED_MARKETS = "/sampling-simplified-markets";
const GET_SAMPLING_MARKETS = "/sampling-markets";
const GET_SIMPLIFIED_MARKETS = "/simplified-markets";
const GET_MARKETS = "/markets";
const GET_MARKET = "/markets/";
const GET_MARKET_BY_TOKEN = "/markets-by-token/";
const GET_CLOB_MARKET = "/clob-markets/";
const GET_ORDER_BOOK = "/book";
const GET_ORDER_BOOKS = "/books";
const GET_MIDPOINT = "/midpoint";
const GET_MIDPOINTS = "/midpoints";
const GET_PRICE = "/price";
const GET_PRICES = "/prices";
const GET_SPREAD = "/spread";
const GET_SPREADS = "/spreads";
const GET_LAST_TRADE_PRICE = "/last-trade-price";
const GET_LAST_TRADES_PRICES = "/last-trades-prices";
const GET_TICK_SIZE = "/tick-size";
const GET_NEG_RISK = "/neg-risk";
const GET_FEE_RATE = "/fee-rate";
const POST_ORDER = "/order";
const POST_ORDERS = "/orders";
const CANCEL_ORDER = "/order";
const CANCEL_ORDERS = "/orders";
const GET_ORDER = "/data/order/";
const CANCEL_ALL = "/cancel-all";
const CANCEL_MARKET_ORDERS = "/cancel-market-orders";
const GET_OPEN_ORDERS = "/data/orders";
const GET_PRE_MIGRATION_ORDERS = "/data/pre-migration-orders";
const GET_TRADES = "/data/trades";
const IS_ORDER_SCORING = "/order-scoring";
const ARE_ORDERS_SCORING = "/orders-scoring";
const GET_PRICES_HISTORY = "/prices-history";
const GET_NOTIFICATIONS = "/notifications";
const DROP_NOTIFICATIONS = "/notifications";
const GET_BALANCE_ALLOWANCE = "/balance-allowance";
const UPDATE_BALANCE_ALLOWANCE = "/balance-allowance/update";
const GET_EARNINGS_FOR_USER_FOR_DAY = "/rewards/user";
const GET_TOTAL_EARNINGS_FOR_USER_FOR_DAY = "/rewards/user/total";
const GET_LIQUIDITY_REWARD_PERCENTAGES = "/rewards/user/percentages";
const GET_REWARDS_MARKETS_CURRENT = "/rewards/markets/current";
const GET_REWARDS_MARKETS = "/rewards/markets/";
const GET_REWARDS_EARNINGS_PERCENTAGES = "/rewards/user/markets";
const CREATE_READONLY_API_KEY = "/auth/readonly-api-key";
const GET_READONLY_API_KEYS = "/auth/readonly-api-keys";
const DELETE_READONLY_API_KEY = "/auth/readonly-api-key";
const CREATE_BUILDER_API_KEY = "/auth/builder-api-key";
const GET_BUILDER_API_KEYS = "/auth/builder-api-key";
const REVOKE_BUILDER_API_KEY = "/auth/builder-api-key";
const GET_MARKET_TRADES_EVENTS = "/markets/live-activity/";
const GET_BUILDER_TRADES = "/builder/trades";
const GET_BUILDER_FEES = "/fees/builder-fees/";

//#endregion
export { ARE_ORDERS_SCORING, CANCEL_ALL, CANCEL_MARKET_ORDERS, CANCEL_ORDER, CANCEL_ORDERS, CLOSED_ONLY, CREATE_API_KEY, CREATE_BUILDER_API_KEY, CREATE_READONLY_API_KEY, DELETE_API_KEY, DELETE_READONLY_API_KEY, DERIVE_API_KEY, DROP_NOTIFICATIONS, GET_API_KEYS, GET_BALANCE_ALLOWANCE, GET_BUILDER_API_KEYS, GET_BUILDER_FEES, GET_BUILDER_TRADES, GET_CLOB_MARKET, GET_EARNINGS_FOR_USER_FOR_DAY, GET_FEE_RATE, GET_LAST_TRADES_PRICES, GET_LAST_TRADE_PRICE, GET_LIQUIDITY_REWARD_PERCENTAGES, GET_MARKET, GET_MARKETS, GET_MARKET_BY_TOKEN, GET_MARKET_TRADES_EVENTS, GET_MIDPOINT, GET_MIDPOINTS, GET_NEG_RISK, GET_NOTIFICATIONS, GET_OPEN_ORDERS, GET_ORDER, GET_ORDER_BOOK, GET_ORDER_BOOKS, GET_PRE_MIGRATION_ORDERS, GET_PRICE, GET_PRICES, GET_PRICES_HISTORY, GET_READONLY_API_KEYS, GET_REWARDS_EARNINGS_PERCENTAGES, GET_REWARDS_MARKETS, GET_REWARDS_MARKETS_CURRENT, GET_SAMPLING_MARKETS, GET_SAMPLING_SIMPLIFIED_MARKETS, GET_SIMPLIFIED_MARKETS, GET_SPREAD, GET_SPREADS, GET_TICK_SIZE, GET_TOTAL_EARNINGS_FOR_USER_FOR_DAY, GET_TRADES, HEARTBEAT, IS_ORDER_SCORING, OK, POST_ORDER, POST_ORDERS, REVOKE_BUILDER_API_KEY, TIME, UPDATE_BALANCE_ALLOWANCE };
