// @polymarket/clob-client-v2/dist/endpoints.cjs

//#region src/endpoints.ts
const OK = "/ok";
const HEARTBEAT = "/v1/heartbeats";
const TIME = "/time";
const CREATE_API_KEY = "/auth/api-key";
const GET_API_KEYS = "/auth/api-keys";
const DELETE_API_KEY = "/auth/api-key";
const DERIVE_API_KEY = "/auth/derive-api-key";
const CLOSED_ONLY = "/auth/ban-status/closed-only";
const GET_SAMPLING_SIMPLIFIED_MARKETS = "/sampling-simplified-markets";
const GET_SAMPLING_MARKETS = "/sampling-markets";
const GET_SIMPLIFIED_MARKETS = "/simplified-markets";
const GET_MARKETS = "/markets";
const GET_MARKET = "/markets/";
const GET_MARKET_BY_TOKEN = "/markets-by-token/";
const GET_CLOB_MARKET = "/clob-markets/";
const GET_ORDER_BOOK = "/book";
const GET_ORDER_BOOKS = "/books";
const GET_MIDPOINT = "/midpoint";
const GET_MIDPOINTS = "/midpoints";
const GET_PRICE = "/price";
const GET_PRICES = "/prices";
const GET_SPREAD = "/spread";
const GET_SPREADS = "/spreads";
const GET_LAST_TRADE_PRICE = "/last-trade-price";
const GET_LAST_TRADES_PRICES = "/last-trades-prices";
const GET_TICK_SIZE = "/tick-size";
const GET_NEG_RISK = "/neg-risk";
const GET_FEE_RATE = "/fee-rate";
const POST_ORDER = "/order";
const POST_ORDERS = "/orders";
const CANCEL_ORDER = "/order";
const CANCEL_ORDERS = "/orders";
const GET_ORDER = "/data/order/";
const CANCEL_ALL = "/cancel-all";
const CANCEL_MARKET_ORDERS = "/cancel-market-orders";
const GET_OPEN_ORDERS = "/data/orders";
const GET_PRE_MIGRATION_ORDERS = "/data/pre-migration-orders";
const GET_TRADES = "/data/trades";
const IS_ORDER_SCORING = "/order-scoring";
const ARE_ORDERS_SCORING = "/orders-scoring";
const GET_PRICES_HISTORY = "/prices-history";
const GET_NOTIFICATIONS = "/notifications";
const DROP_NOTIFICATIONS = "/notifications";
const GET_BALANCE_ALLOWANCE = "/balance-allowance";
const UPDATE_BALANCE_ALLOWANCE = "/balance-allowance/update";
const GET_EARNINGS_FOR_USER_FOR_DAY = "/rewards/user";
const GET_TOTAL_EARNINGS_FOR_USER_FOR_DAY = "/rewards/user/total";
const GET_LIQUIDITY_REWARD_PERCENTAGES = "/rewards/user/percentages";
const GET_REWARDS_MARKETS_CURRENT = "/rewards/markets/current";
const GET_REWARDS_MARKETS = "/rewards/markets/";
const GET_REWARDS_EARNINGS_PERCENTAGES = "/rewards/user/markets";
const CREATE_READONLY_API_KEY = "/auth/readonly-api-key";
const GET_READONLY_API_KEYS = "/auth/readonly-api-keys";
const DELETE_READONLY_API_KEY = "/auth/readonly-api-key";
const CREATE_BUILDER_API_KEY = "/auth/builder-api-key";
const GET_BUILDER_API_KEYS = "/auth/builder-api-key";
const REVOKE_BUILDER_API_KEY = "/auth/builder-api-key";
const GET_MARKET_TRADES_EVENTS = "/markets/live-activity/";
const GET_BUILDER_TRADES = "/builder/trades";
const GET_BUILDER_FEES = "/fees/builder-fees/";

//#endregion
exports.ARE_ORDERS_SCORING = ARE_ORDERS_SCORING;
exports.CANCEL_ALL = CANCEL_ALL;
exports.CANCEL_MARKET_ORDERS = CANCEL_MARKET_ORDERS;
exports.CANCEL_ORDER = CANCEL_ORDER;
exports.CANCEL_ORDERS = CANCEL_ORDERS;
exports.CLOSED_ONLY = CLOSED_ONLY;
exports.CREATE_API_KEY = CREATE_API_KEY;
exports.CREATE_BUILDER_API_KEY = CREATE_BUILDER_API_KEY;
exports.CREATE_READONLY_API_KEY = CREATE_READONLY_API_KEY;
exports.DELETE_API_KEY = DELETE_API_KEY;
exports.DELETE_READONLY_API_KEY = DELETE_READONLY_API_KEY;
exports.DERIVE_API_KEY = DERIVE_API_KEY;
exports.DROP_NOTIFICATIONS = DROP_NOTIFICATIONS;
exports.GET_API_KEYS = GET_API_KEYS;
exports.GET_BALANCE_ALLOWANCE = GET_BALANCE_ALLOWANCE;
exports.GET_BUILDER_API_KEYS = GET_BUILDER_API_KEYS;
exports.GET_BUILDER_FEES = GET_BUILDER_FEES;
exports.GET_BUILDER_TRADES = GET_BUILDER_TRADES;
exports.GET_CLOB_MARKET = GET_CLOB_MARKET;
exports.GET_EARNINGS_FOR_USER_FOR_DAY = GET_EARNINGS_FOR_USER_FOR_DAY;
exports.GET_FEE_RATE = GET_FEE_RATE;
exports.GET_LAST_TRADES_PRICES = GET_LAST_TRADES_PRICES;
exports.GET_LAST_TRADE_PRICE = GET_LAST_TRADE_PRICE;
exports.GET_LIQUIDITY_REWARD_PERCENTAGES = GET_LIQUIDITY_REWARD_PERCENTAGES;
exports.GET_MARKET = GET_MARKET;
exports.GET_MARKETS = GET_MARKETS;
exports.GET_MARKET_BY_TOKEN = GET_MARKET_BY_TOKEN;
exports.GET_MARKET_TRADES_EVENTS = GET_MARKET_TRADES_EVENTS;
exports.GET_MIDPOINT = GET_MIDPOINT;
exports.GET_MIDPOINTS = GET_MIDPOINTS;
exports.GET_NEG_RISK = GET_NEG_RISK;
exports.GET_NOTIFICATIONS = GET_NOTIFICATIONS;
exports.GET_OPEN_ORDERS = GET_OPEN_ORDERS;
exports.GET_ORDER = GET_ORDER;
exports.GET_ORDER_BOOK = GET_ORDER_BOOK;
exports.GET_ORDER_BOOKS = GET_ORDER_BOOKS;
exports.GET_PRE_MIGRATION_ORDERS = GET_PRE_MIGRATION_ORDERS;
exports.GET_PRICE = GET_PRICE;
exports.GET_PRICES = GET_PRICES;
exports.GET_PRICES_HISTORY = GET_PRICES_HISTORY;
exports.GET_READONLY_API_KEYS = GET_READONLY_API_KEYS;
exports.GET_REWARDS_EARNINGS_PERCENTAGES = GET_REWARDS_EARNINGS_PERCENTAGES;
exports.GET_REWARDS_MARKETS = GET_REWARDS_MARKETS;
exports.GET_REWARDS_MARKETS_CURRENT = GET_REWARDS_MARKETS_CURRENT;
exports.GET_SAMPLING_MARKETS = GET_SAMPLING_MARKETS;
exports.GET_SAMPLING_SIMPLIFIED_MARKETS = GET_SAMPLING_SIMPLIFIED_MARKETS;
exports.GET_SIMPLIFIED_MARKETS = GET_SIMPLIFIED_MARKETS;
exports.GET_SPREAD = GET_SPREAD;
exports.GET_SPREADS = GET_SPREADS;
exports.GET_TICK_SIZE = GET_TICK_SIZE;
exports.GET_TOTAL_EARNINGS_FOR_USER_FOR_DAY = GET_TOTAL_EARNINGS_FOR_USER_FOR_DAY;
exports.GET_TRADES = GET_TRADES;
exports.HEARTBEAT = HEARTBEAT;
exports.IS_ORDER_SCORING = IS_ORDER_SCORING;
exports.OK = OK;
exports.POST_ORDER = POST_ORDER;
exports.POST_ORDERS = POST_ORDERS;
exports.REVOKE_BUILDER_API_KEY = REVOKE_BUILDER_API_KEY;
exports.TIME = TIME;
exports.UPDATE_BALANCE_ALLOWANCE = UPDATE_BALANCE_ALLOWANCE;
