// @polymarket/clob-client-v2/dist/index.js
import { ApiError, L1_AUTH_UNAVAILABLE_ERROR, L2_AUTH_NOT_AVAILABLE } from "./errors.js";
import { adjustBuyAmountForFees } from "./fees/index.js";
import { createL1Headers, createL2Headers } from "./headers/index.js";
import { COLLATERAL_TOKEN_DECIMALS, CONDITIONAL_TOKEN_DECIMALS, getContractConfig } from "./config.js";
import { SignatureTypeV1 } from "./order-utils/model/signatureTypeV1.js";
import { SignatureTypeV2 } from "./order-utils/model/signatureTypeV2.js";
import { Side } from "./order-utils/model/side.js";
import { AssetType, Chain, OrderType, PriceHistoryInterval } from "./types/clob.js";
import { orderToJsonV1 } from "./types/ordersV1.js";
import { orderToJsonV2 } from "./types/ordersV2.js";
import { isV2Order } from "./types/unifiedOrder.js";
import { OrderBuilder } from "./order-builder/orderBuilder.js";
import { ClobClient } from "./client.js";

export { ApiError, AssetType, COLLATERAL_TOKEN_DECIMALS, CONDITIONAL_TOKEN_DECIMALS, Chain, ClobClient, L1_AUTH_UNAVAILABLE_ERROR, L2_AUTH_NOT_AVAILABLE, OrderBuilder, OrderType, PriceHistoryInterval, Side, SignatureTypeV1, SignatureTypeV2, adjustBuyAmountForFees, createL1Headers, createL2Headers, getContractConfig, isV2Order, orderToJsonV1, orderToJsonV2 };
