// @polymarket/clob-client-v2/dist/index.cjs
const require_errors = require('./errors.cjs');
const require_index = require('./fees/index.cjs');
const require_index$1 = require('./headers/index.cjs');
const require_config = require('./config.cjs');
const require_signatureTypeV1 = require('./order-utils/model/signatureTypeV1.cjs');
const require_signatureTypeV2 = require('./order-utils/model/signatureTypeV2.cjs');
const require_side = require('./order-utils/model/side.cjs');
const require_clob = require('./types/clob.cjs');
const require_ordersV1 = require('./types/ordersV1.cjs');
const require_ordersV2 = require('./types/ordersV2.cjs');
const require_unifiedOrder = require('./types/unifiedOrder.cjs');
const require_orderBuilder = require('./order-builder/orderBuilder.cjs');
const require_client = require('./client.cjs');

exports.ApiError = require_errors.ApiError;
exports.AssetType = require_clob.AssetType;
exports.COLLATERAL_TOKEN_DECIMALS = require_config.COLLATERAL_TOKEN_DECIMALS;
exports.CONDITIONAL_TOKEN_DECIMALS = require_config.CONDITIONAL_TOKEN_DECIMALS;
exports.Chain = require_clob.Chain;
exports.ClobClient = require_client.ClobClient;
exports.L1_AUTH_UNAVAILABLE_ERROR = require_errors.L1_AUTH_UNAVAILABLE_ERROR;
exports.L2_AUTH_NOT_AVAILABLE = require_errors.L2_AUTH_NOT_AVAILABLE;
exports.OrderBuilder = require_orderBuilder.OrderBuilder;
exports.OrderType = require_clob.OrderType;
exports.PriceHistoryInterval = require_clob.PriceHistoryInterval;
exports.Side = require_side.Side;
exports.SignatureTypeV1 = require_signatureTypeV1.SignatureTypeV1;
exports.SignatureTypeV2 = require_signatureTypeV2.SignatureTypeV2;
exports.adjustBuyAmountForFees = require_index.adjustBuyAmountForFees;
exports.createL1Headers = require_index$1.createL1Headers;
exports.createL2Headers = require_index$1.createL2Headers;
exports.getContractConfig = require_config.getContractConfig;
exports.isV2Order = require_unifiedOrder.isV2Order;
exports.orderToJsonV1 = require_ordersV1.orderToJsonV1;
exports.orderToJsonV2 = require_ordersV2.orderToJsonV2;
