// @polymarket/clob-client-v2/dist/utilities.cjs

//#region src/utilities.ts
const roundNormal = (num, decimals) => {
	if (decimalPlaces(num) <= decimals) return num;
	return Math.round((num + Number.EPSILON) * 10 ** decimals) / 10 ** decimals;
};
const roundDown = (num, decimals) => {
	if (decimalPlaces(num) <= decimals) return num;
	return Math.floor(num * 10 ** decimals) / 10 ** decimals;
};
const roundUp = (num, decimals) => {
	if (decimalPlaces(num) <= decimals) return num;
	return Math.ceil(num * 10 ** decimals) / 10 ** decimals;
};
const decimalPlaces = (num) => {
	if (Number.isInteger(num)) return 0;
	const arr = num.toString().split(".");
	if (arr.length <= 1) return 0;
	return arr[1].length;
};
/**
* Calculates the hash for the given orderbook
* @param orderbook
* @returns
*/
const generateOrderBookSummaryHash = async (orderbook) => {
	if (!globalThis.crypto?.subtle) throw new Error("generateOrderBookSummaryHash: globalThis.crypto.subtle is unavailable. Requires Node >=20 or a secure browser context (HTTPS or localhost).");
	orderbook.hash = "";
	const data = new TextEncoder().encode(JSON.stringify(orderbook));
	const hashBuffer = await globalThis.crypto.subtle.digest("SHA-1", data);
	const hash = Array.from(new Uint8Array(hashBuffer)).map((b) => b.toString(16).padStart(2, "0")).join("");
	orderbook.hash = hash;
	return hash;
};
const isTickSizeSmaller = (a, b) => {
	return parseFloat(a) < parseFloat(b);
};
const priceValid = (price, tickSize) => {
	return price >= parseFloat(tickSize) && price <= 1 - parseFloat(tickSize);
};

//#endregion
exports.decimalPlaces = decimalPlaces;
exports.generateOrderBookSummaryHash = generateOrderBookSummaryHash;
exports.isTickSizeSmaller = isTickSizeSmaller;
exports.priceValid = priceValid;
exports.roundDown = roundDown;
exports.roundNormal = roundNormal;
exports.roundUp = roundUp;
