// @polymarket/clob-client-v2/dist/headers/index.js
import { getSignerAddress } from "../signing/signer.js";
import { buildClobEip712Signature } from "../signing/eip712.js";
import { buildPolyHmacSignature } from "../signing/hmac.js";

//#region src/headers/index.ts
const createL1Headers = async (signer, chainId, nonce, timestamp, address) => {
	let ts = Math.floor(Date.now() / 1e3);
	if (timestamp !== void 0) ts = timestamp;
	let n = 0;
	if (nonce !== void 0) n = nonce;
	const sig = await buildClobEip712Signature(signer, chainId, ts, n, address);
	return {
		POLY_ADDRESS: address ?? await getSignerAddress(signer),
		POLY_SIGNATURE: sig,
		POLY_TIMESTAMP: `${ts}`,
		POLY_NONCE: `${n}`
	};
};
const createL2Headers = async (signer, creds, l2HeaderArgs, timestamp) => {
	let ts = Math.floor(Date.now() / 1e3);
	if (timestamp !== void 0) ts = timestamp;
	return {
		POLY_ADDRESS: await getSignerAddress(signer),
		POLY_SIGNATURE: await buildPolyHmacSignature(creds.secret, ts, l2HeaderArgs.method, l2HeaderArgs.requestPath, l2HeaderArgs.body),
		POLY_TIMESTAMP: `${ts}`,
		POLY_API_KEY: creds.key,
		POLY_PASSPHRASE: creds.passphrase
	};
};

//#endregion
export { createL1Headers, createL2Headers };
