// @polymarket/clob-client-v2/dist/headers/index.cjs
const require_signer = require('../signing/signer.cjs');
const require_eip712 = require('../signing/eip712.cjs');
const require_hmac = require('../signing/hmac.cjs');

//#region src/headers/index.ts
const createL1Headers = async (signer, chainId, nonce, timestamp, address) => {
	let ts = Math.floor(Date.now() / 1e3);
	if (timestamp !== void 0) ts = timestamp;
	let n = 0;
	if (nonce !== void 0) n = nonce;
	const sig = await require_eip712.buildClobEip712Signature(signer, chainId, ts, n, address);
	return {
		POLY_ADDRESS: address ?? await require_signer.getSignerAddress(signer),
		POLY_SIGNATURE: sig,
		POLY_TIMESTAMP: `${ts}`,
		POLY_NONCE: `${n}`
	};
};
const createL2Headers = async (signer, creds, l2HeaderArgs, timestamp) => {
	let ts = Math.floor(Date.now() / 1e3);
	if (timestamp !== void 0) ts = timestamp;
	return {
		POLY_ADDRESS: await require_signer.getSignerAddress(signer),
		POLY_SIGNATURE: await require_hmac.buildPolyHmacSignature(creds.secret, ts, l2HeaderArgs.method, l2HeaderArgs.requestPath, l2HeaderArgs.body),
		POLY_TIMESTAMP: `${ts}`,
		POLY_API_KEY: creds.key,
		POLY_PASSPHRASE: creds.passphrase
	};
};

//#endregion
exports.createL1Headers = createL1Headers;
exports.createL2Headers = createL2Headers;
