// @polymarket/clob-client-v2/dist/order-utils/exchangeOrderBuilderV1.cjs
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.cjs');
const require_signer = require('../signing/signer.cjs');
const require_ctfExchangeV1TypedData = require('./model/ctfExchangeV1TypedData.cjs');
const require_eip712 = require('./model/eip712.cjs');
const require_signatureTypeV1 = require('./model/signatureTypeV1.cjs');
const require_utils = require('./utils.cjs');
let viem = require("viem");

//#region src/order-utils/exchangeOrderBuilderV1.ts
const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
var ExchangeOrderBuilderV1 = class {
	constructor(contractAddress, chainId, signer, generateSalt = require_utils.generateOrderSalt) {
		this.contractAddress = contractAddress;
		this.chainId = chainId;
		this.signer = signer;
		this.generateSalt = generateSalt;
	}
	/**
	* build an order object including the signature.
	* @param orderData
	* @returns a SignedOrder object (order + signature)
	*/
	async buildSignedOrder(orderData) {
		const order = await this.buildOrder(orderData);
		const orderTypedData = this.buildOrderTypedData(order);
		const orderSignature = await this.buildOrderSignature(orderTypedData);
		return {
			...order,
			signature: orderSignature
		};
	}
	/**
	* Creates an Order object from order data.
	* @param OrderData
	* @returns a Order object (not signed)
	*/
	async buildOrder({ maker, taker, tokenId, makerAmount, takerAmount, side, feeRateBps, nonce, signer, expiration, signatureType }) {
		if (!signer) signer = maker;
		const signerAddress = await require_signer.getSignerAddress(this.signer);
		if (signer !== signerAddress) throw new Error("signer does not match");
		if (!expiration) expiration = "0";
		if (!signatureType) signatureType = require_signatureTypeV1.SignatureTypeV1.EOA;
		let takerAddress;
		if (typeof taker !== "undefined" && taker) takerAddress = taker;
		else takerAddress = ZERO_ADDRESS;
		let feeRateBpsResolved;
		if (typeof feeRateBps !== "undefined" && feeRateBps) feeRateBpsResolved = feeRateBps.toString();
		else feeRateBpsResolved = "0";
		let nonceResolved;
		if (typeof nonce !== "undefined" && nonce) nonceResolved = nonce.toString();
		else nonceResolved = "0";
		return {
			salt: this.generateSalt(),
			maker,
			signer,
			taker: takerAddress,
			tokenId,
			makerAmount,
			takerAmount,
			expiration,
			nonce: nonceResolved,
			feeRateBps: feeRateBpsResolved,
			side,
			signatureType
		};
	}
	/**
	* Parses an Order object to EIP712 typed data
	* @param order
	* @returns a EIP712TypedData object
	*/
	buildOrderTypedData(order) {
		return {
			primaryType: "Order",
			types: {
				EIP712Domain: require_eip712.EIP712_DOMAIN,
				Order: require_ctfExchangeV1TypedData.CTF_EXCHANGE_V1_ORDER_STRUCT
			},
			domain: {
				name: require_ctfExchangeV1TypedData.CTF_EXCHANGE_V1_DOMAIN_NAME,
				version: require_ctfExchangeV1TypedData.CTF_EXCHANGE_V1_DOMAIN_VERSION,
				chainId: this.chainId,
				verifyingContract: this.contractAddress
			},
			message: {
				salt: order.salt,
				maker: order.maker,
				signer: order.signer,
				taker: order.taker,
				tokenId: order.tokenId,
				makerAmount: order.makerAmount,
				takerAmount: order.takerAmount,
				expiration: order.expiration,
				nonce: order.nonce,
				feeRateBps: order.feeRateBps,
				side: order.side === "BUY" ? 0 : 1,
				signatureType: order.signatureType
			}
		};
	}
	/**
	* Generates order's signature from a EIP712TypedData object + the signer address
	* @param typedData
	* @returns a OrderSignature that is an string
	*/
	buildOrderSignature(typedData) {
		delete typedData.types.EIP712Domain;
		return require_signer.signTypedDataWithSigner({
			signer: this.signer,
			domain: typedData.domain,
			types: typedData.types,
			value: typedData.message,
			primaryType: typedData.primaryType
		});
	}
	/**
	* Generates the hash of the order from a EIP712TypedData object.
	* @param orderTypedData
	* @returns a OrderHash that is an string
	*/
	buildOrderHash(orderTypedData) {
		return (0, viem.hashTypedData)(orderTypedData);
	}
};

//#endregion
exports.ExchangeOrderBuilderV1 = ExchangeOrderBuilderV1;
